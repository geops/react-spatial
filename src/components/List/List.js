import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import ListItem from '../ListItem';

const propTypes = {
  className: PropTypes.string,
  items: PropTypes.array,
  renderItem: PropTypes.func,
  onSelect: PropTypes.func,
  onKeyDownItem: PropTypes.func,
  getItemKey: PropTypes.func.isRequired,
};

const defaultProps = {
  className: 'tm-list',
  items: [],
  renderItem: () => {},
  onSelect: () => {},
  onKeyDownItem: () => {},
};

/**
 * This component displays a `<ul>` HMTL tag element.
 *
 * This component also add keyboard navigation (arrow up/down, w/s) between list
 * items.
 */
class List extends PureComponent {
  render() {
    const {
      className,
      items,
      renderItem,
      onSelect,
      onKeyDownItem,
      getItemKey,
    } = this.props;

    if (!items.length) {
      return null;
    }

    return (
      <ul role="menu" className={className}>
        {items.map(item => (
          <ListItem
            key={getItemKey(item)}
            item={item}
            onSelect={(e, itm) => {
              onSelect(e, itm);
            }}
            onKeyDown={e => {
              onKeyDownItem(e);
            }}
          >
            {renderItem(item)}
          </ListItem>
        ))}
      </ul>
    );
  }
}

List.propTypes = propTypes;
List.defaultProps = defaultProps;

export default List;
