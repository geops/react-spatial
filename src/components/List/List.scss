.tm-list {
  list-style-type: none;
  padding: 0;
  margin: 0;

  li {
    line-height: 25px;
    padding: 4px 15px;
    cursor: pointer;
  }
}
