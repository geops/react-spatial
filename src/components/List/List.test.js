import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import List from './List';

configure({ adapter: new Adapter() });

describe('List', () => {
  describe('when no properties are set', () => {
    let spy = null;

    beforeEach(() => {
      window.console.error = jest.fn().mockImplementation(() => {});
      spy = jest.spyOn(window.console, 'error');
    });

    afterEach(() => {
      spy.mockRestore();
      window.console.error.mockRestore();
    });

    test('displays 1 error for required property ', () => {
      shallow(<List />);
      expect(spy).toHaveBeenCalledTimes(1);
    });

    test('matches snapshot', () => {
      const component = renderer.create(<List />);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when properties are set', () => {
    const defaultItems = [
      {
        label: 'foo',
      },
      {
        label: 'bar',
      },
      {
        label: 'foo2',
      },
    ];

    const items = [
      {
        label: 'qux',
      },
      {
        label: 'quux',
      },
      {
        label: 'corge',
      },
    ];

    test('matches snapshot with defaultItems', () => {
      const component = renderer.create(
        <List
          className="tm-foo"
          defaultItems={defaultItems}
          renderTitle={() => 'my_foo_title'}
          renderItem={item => item.label}
          getItemKey={() => Math.random()}
          onSelect={() => {}}
        />,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('matches snapshot with items', () => {
      const component = renderer.create(
        <List
          className="tm-foo"
          defaultItems={defaultItems}
          items={items}
          renderTitle={() => 'my_foo_title'}
          renderItem={item => item.label}
          getItemKey={() => Math.random()}
          onSelect={() => {}}
        />,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
