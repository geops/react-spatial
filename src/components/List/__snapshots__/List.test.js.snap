// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List when no properties are set matches snapshot 1`] = `null`;

exports[`List when properties are set matches snapshot with defaultItems 1`] = `null`;

exports[`List when properties are set matches snapshot with items 1`] = `
<ul
  className="tm-foo"
  role="menu"
>
  <li
    className="tm-list-item"
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    role="menuitem"
    tabIndex="0"
  >
    qux
  </li>
  <li
    className="tm-list-item"
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    role="menuitem"
    tabIndex="0"
  >
    quux
  </li>
  <li
    className="tm-list-item"
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    role="menuitem"
    tabIndex="0"
  >
    corge
  </li>
</ul>
`;
